### PR TITLE
disttask: add load data dispatcher

### LINKED ISSUE #42591: add dispatcher for load data

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: close #42591

Problem Summary:

### What is changed and how it works?
- split files into chunks
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/loaddata/BUILD.bazel [hunk 1]
```diff
@@ -1,15 +1,40 @@
...
         "//parser/mysql",
+        "//util/intest",
+        "//util/logutil",
+        "@com_github_pingcap_errors//:errors",
+        "@org_uber_go_zap//:zap",
+    ],
+)
+
+go_test(
+    name = "loaddata_test",
```
Comment (hawkingrei) [inline disttask/loaddata/BUILD.bazel:30]: Please add 

```
timeout = "short",
flaky = True,
```

File: disttask/loaddata/dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			EndOffset:    region.Chunk.EndOffset,
+			RealOffset:   region.Chunk.RealOffset,
+			PrevRowIDMax: region.Chunk.PrevRowIDMax,
+			RowIDMax:     region.Chunk.RowIDMax,
+		})
+	}
+	return subtasks, nil
+}
+
+func init() {
```
Comment (hawkingrei) [inline disttask/loaddata/dispatcher.go:105]: We should avoid ```init```. https://github.com/uber-go/guide/blob/master/style.md#avoid-init

File: disttask/framework/dispatcher/dispatcher.go [hunk 3]
```diff
@@ -50,6 +50,12 @@ type Dispatch interface {
 	Stop()
 }
 
+// Handle provides the interface for operations needed by task flow handles.
+type Handle interface {
+	// GetTaskAllInstances gets handles the task's all available instances.
+	GetTaskAllInstances(ctx context.Context, gTaskID int64) ([]string, error)
```
Comment (lance6716) [inline disttask/framework/dispatcher/dispatcher.go:56]: > gTaskID

Handle is per task/dispatcher node/cluster?

File: disttask/framework/dispatcher/dispatcher_mock.go
Comment (lance6716) [inline disttask/framework/dispatcher/dispatcher_mock.go:1]: in https://github.com/stretchr/testify#mock-package

> You can use the [mockery tool](http://github.com/vektra/mockery) to autogenerate the mock code against an interface as well, making using mocks much quicker.

Can we add a makefile entry for it?

File: disttask/framework/dispatcher/register.go [hunk 4]
```diff
@@ -15,14 +15,16 @@
...
+
 	"github.com/pingcap/tidb/disttask/framework/proto"
 	"github.com/pingcap/tidb/util/syncutil"
 )
 
 // TaskFlowHandle is used to control the process operations for each global task.
 type TaskFlowHandle interface {
-	ProcessNormalFlow(d Dispatch, gTask *proto.Task) (metas [][]byte, err error)
-	ProcessErrFlow(d Dispatch, gTask *proto.Task, receive string) (meta []byte, err error)
+	ProcessNormalFlow(ctx context.Context, h Handle, gTask *proto.Task) (metas [][]byte, err error)
```
Comment (lance6716) [inline disttask/framework/dispatcher/register.go:26]: the comment says this is for each task, so why we need to specify `gTask` as argument?

File: disttask/loaddata/dispatcher.go (hunk 2 above)
Comment (GMHDBJD) [reply disttask/loaddata/dispatcher.go:105]: I think the init here is inevitable and meets the 4 guidelines in the documentation.
> 1. Be completely deterministic, regardless of program environment or invocation.
> 2. Avoid depending on the ordering or side-effects of other init() functions. While init() ordering is well-known, code can change, and thus relationships between init() functions can make code brittle and error-prone.
> 3. Avoid accessing or manipulating global or environment state, such as machine information, environment variables, working directory, program arguments/inputs, etc.
> 4. Avoid I/O, including both filesystem, network, and system calls.

File: disttask/framework/dispatcher/dispatcher_mock.go
Comment (GMHDBJD) [reply disttask/framework/dispatcher/dispatcher_mock.go:1]: I can add in later pr with https://github.com/pingcap/tidb/blob/master/disttask/framework/scheduler/interface_mock.go

File: disttask/framework/dispatcher/dispatcher.go (hunk 3 above)
Comment (GMHDBJD) [reply disttask/framework/dispatcher/dispatcher.go:56]: per dispatcher

File: disttask/loaddata/dispatcher.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	if err != nil {
+		return nil, err
+	}
+	logutil.BgLogger().Info("process normal flow", zap.Any("task_meta", taskMeta), zap.Any("step", gTask.Step))
+
+	switch gTask.Step {
+	case Import:
+		gTask.State = proto.TaskStateSucceed
+		return nil, nil
+	default:
```
Comment (lance6716) [inline disttask/loaddata/dispatcher.go:43]: it's very ambiguous to use a value out of enum value set. Can you add a default step?

File: disttask/loaddata/dispatcher.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	taskMeta := &TaskMeta{}
+	err := json.Unmarshal(gTask.Meta, taskMeta)
+	if err != nil {
+		return nil, err
+	}
+	logutil.BgLogger().Info("process normal flow", zap.Any("task_meta", taskMeta), zap.Any("step", gTask.Step))
+
+	switch gTask.Step {
+	case Import:
+		gTask.State = proto.TaskStateSucceed
```
Comment (lance6716) [inline disttask/loaddata/dispatcher.go:41]: What's the difference of `Step` and `State`? Please add comments

File: disttask/loaddata/dispatcher.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	logutil.BgLogger().Info("generate subtasks", zap.Any("subtask_metas", subtaskMetas))
+	metaBytes := make([][]byte, 0, len(taskMeta.FileInfos))
+	for _, subtaskMeta := range subtaskMetas {
+		bs, err := json.Marshal(subtaskMeta)
+		if err != nil {
+			return nil, err
+		}
+		metaBytes = append(metaBytes, bs)
+	}
+	gTask.Step = Import
```
Comment (lance6716) [inline disttask/loaddata/dispatcher.go:63]: please add comment to the interface that `gTask` will be modified inside this method

File: disttask/loaddata/dispatcher.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+func generateSubtaskMetas(ctx context.Context, task *TaskMeta, concurrency int) ([]*SubtaskMeta, error) {
+	tableRegions, err := makeTableRegions(ctx, task, concurrency)
+	if err != nil {
+		return nil, err
+	}
+
+	engineMap := make(map[int32]int)
+	subtasks := make([]*SubtaskMeta, 0)
+	for _, region := range tableRegions {
```
Comment (lance6716) [inline disttask/loaddata/dispatcher.go:81]: more readable that use map{EngineID: SubtaskMeta} and https://pkg.go.dev/golang.org/x/exp/maps#Values

File: disttask/loaddata/dispatcher.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+			RealOffset:   region.Chunk.RealOffset,
+			PrevRowIDMax: region.Chunk.PrevRowIDMax,
+			RowIDMax:     region.Chunk.RowIDMax,
+		})
+	}
+	return subtasks, nil
+}
+
+func init() {
+	dispatcher.RegisterTaskFlowHandle(proto.LoadData, &FlowHandle{})
```
Comment (lance6716) [inline disttask/loaddata/dispatcher.go:106]: no need to use lock inside RegisterTaskFlowHandle?

https://go.dev/ref/spec

> Package initialization—variable initialization and the invocation of init functions—happens in a single goroutine, sequentially, one package at a time.

File: disttask/loaddata/dispatcher.go (hunk 6 above)
Comment (GMHDBJD) [reply disttask/loaddata/dispatcher.go:41]: https://github.com/pingcap/tidb/blob/master/disttask/framework/proto/task.go#L21-L51

File: disttask/loaddata/dispatcher.go (hunk 2 above)
Comment (hawkingrei) [reply disttask/loaddata/dispatcher.go:103]: OK

File: disttask/framework/dispatcher/register.go (hunk 4 above)
Comment (lance6716) [reply disttask/framework/dispatcher/register.go:26]: PTAL @zimulala , maybe let `RegisterTaskFlowHandle` register a factory function

File: disttask/loaddata/dispatcher.go (hunk 7 above)
Comment (lance6716) [reply disttask/loaddata/dispatcher.go:64]: @zimulala Not sure if you can fix it in your PR conveniently

File: disttask/loaddata/dispatcher.go (hunk 9 above)
Comment (lance6716) [reply disttask/loaddata/dispatcher.go:104]: @zimulala

**Review by lance6716 — APPROVED** [review]

File: disttask/framework/dispatcher/dispatcher.go [hunk 10]
```diff
@@ -50,6 +50,12 @@ type Dispatch interface {
 	Stop()
 }
 
+// Handle provides the interface for operations needed by task flow handles.
+type Handle interface {
```
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher.go:54]:
```suggestion
type TaskHandle interface {
```

File: disttask/framework/dispatcher/dispatcher.go (hunk 3 above)
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher.go:56]: `GetAllSchedulerIDs`?

File: disttask/framework/dispatcher/dispatcher_mock.go
Comment (D3Hunter) [inline disttask/framework/dispatcher/dispatcher_mock.go:1]: generate it with `mockgen`? and put it into Makefile

Comment (GMHDBJD) [reply disttask/framework/dispatcher/dispatcher_mock.go:1]: ok, in next pr https://github.com/pingcap/tidb/pull/42592#discussion_r1150078455

File: disttask/framework/dispatcher/dispatcher.go (hunk 3 above)
Comment (GMHDBJD) [reply disttask/framework/dispatcher/dispatcher.go]: done

File: disttask/framework/dispatcher/BUILD.bazel [hunk 11]
```diff
@@ -16,6 +17,7 @@ go_library(
         "//util/logutil",
         "//util/syncutil",
         "@com_github_pingcap_errors//:errors",
+        "@com_github_stretchr_testify//mock",
```
Comment (hawkingrei) [inline disttask/framework/dispatcher/BUILD.bazel:20]: I suggest we can add ```race = "on",``` to ```go_test``` for finding the data race. You can decide whether to open it according to the situation

**Review by D3Hunter — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: disttask/loaddata/dispatcher.go (hunk 7 above)
Comment (zimulala) [reply disttask/loaddata/dispatcher.go:64]: I'll comment on that

File: disttask/loaddata/dispatcher.go (hunk 9 above)
Comment (zimulala) [reply disttask/loaddata/dispatcher.go:104]: But we're not sure that all the calls are inside the init function

**Pull request merged**